### ISSUE TITLE
Child Windows under Wayland are clipped on resizing or on showing when placed partially outside the main window

### ISSUE THREAD

**Issue by ggarra13** [fltk/fltk#987]
**Describe the bug**
Wayland supports subwindows extending outside the boundaries of the main window.
However, FLTK clips and redraws them incorrectly when:

1) The subwindow is shown and it is partially outside its main parent window.  It the subwindow is completely outside or completely inside the main window, there's no clipping.
2) The main window is resized and the subwindow is made to overlap.  Leads to redraw issues.
3) The subwindow is resized and it is made to overlap the main window.  Leads to redraw issues.

**To Reproduce**
See Issue #983 for a test program and for more information.

**Expected behavior**
Clipping should not happen as this is not a limitation of Wayland.  Since Wayland's top windows have limited functionality it is imperative to support subwindows fully.

**Linux/Unix Runtime, if applicable:**
 - Wayland

**Comment by ManoloFLTK** [comment]
The attached patch adds member function Fl_Window::can_leak_outside_parent(bool) to remove the clipping of subwindows. Does it fix the issue?

[issue987.patch](https://github.com/user-attachments/files/15522572/issue987.patch)

**Comment by ggarra13** [comment]
Excellent work, Manolo!  It works flawlessly!   Thank you very mucho!  I noticed you also changed it for macOS (Fl_cocoa.mm), but I did not test it.  Let me know if you want me to test it there too.

**Comment by ManoloFLTK** [comment]
OK. I have committed a slightly different patch that is effective only under Wayland and has this public member function: `Fl_Window::allow_expand_outside_parent()`.

**Comment by ManoloFLTK** [comment]
Gonzalo: when using your test program, I find that the "Drag me" window wobbles a lot when it gets dragged. Rather than looking in detail at your dragging code, I have prepared a test program that allows to drag a subwindow around with the mouse. That's a small modification of FLTK's subwindow program (attached patch). With this way of following the mouse, I see no wobbling at all.

[subwindow.patch](https://github.com/user-attachments/files/15524508/subwindow.patch)

**Comment by ggarra13** [comment]
I did not get to update to the latest code so that's why you may see wobbling (also I was using a timeout which is not ideal and might need different values depending on the speed of your machine).

I see what you did (adding init_sizes, which might be the ticket I was missing).   

The reason I went with a timeout was that I found that dragging and resizing heavy windows full of widgets would sometimes give really bad numbers (like -2455567) from event_x_root()/event_y_root().

UPDATE:  I just tried the subwindow example and it happens that for me too.  The subwindow goes flying into space as soon as I drag it on my Ubuntu 22.04.4 LTS box with the Noble libnvidia-egl-wayland1 drivers.

**Comment by ggarra13** [comment]
I found the timeout was needed as it seems the compositor/window manager gets into a crazy loop when you modify it from FL_EVENT.

**Comment by ggarra13** [comment]
Here's an update to my code to with timeouts and init_sizes on the timeout. Tell me if you see wobbling on your side.

[wayland_dragsubwindow_bug_v1.zip](https://github.com/user-attachments/files/15524705/wayland_child_window_bug.zip)

**Comment by ManoloFLTK** [comment]
I am not going to work with your code. It's too complex. We need a small program focussed only on the problem under consideration, moving a child window. subwindow has this property.

I believe the proper solution to the wobbling is different and does not involve a timeout with an arbitrary delay value. 

What do you see in terms of wobbling with the subwindow program, the last source code of FLTK and the mesa driver ?

**Comment by ggarra13** [comment]
Well, my code actually works on my machine.  Here's what your subwindow code does on mine.  I just added some printouts.  I attach your patched source code and the log of the offsets I get:

[subwindow.zip](https://github.com/user-attachments/files/15524751/subwindow.zip)

**Comment by ggarra13** [comment]
> What do you see in terms of wobbling with the subwindow program, the last source code of FLTK and the mesa driver ?

I get no wobbling, just a subwindow that flies out into space and disappears from the screen.

**Comment by ggarra13** [comment]
I attach the log for the same code running on X11, which obviously works, but clips at the edge of the window.

[log.txt](https://github.com/user-attachments/files/15524836/log.txt)

**Comment by ManoloFLTK** [comment]
I hope the modified subwindow test app should be fixed with this patch:

[issue987V2.patch](https://github.com/user-attachments/files/15526256/issue987V2.patch)

**Comment by ManoloFLTK** [comment]
and this other patch is simpler and should be good too:

[issue987V3.patch](https://github.com/user-attachments/files/15526351/issue987V3.patch)

**Comment by ggarra13** [comment]
Yes.  Seems to work fine in the simple demo.  I am testing it with my code.

**Comment by ggarra13** [comment]
Okay.  I tested with my code with V3 patch.  Dragging works fine.  However, resizing the subwindow does not work properly.  In my program, it stretches the window outside the screen if I drag from either the top or from the left (dragging from the right, bottom or bottom+right corner works fine).  If dragging too far, it will lead to (window outside the screen and this error):

Can't create Cairo surface with cairo_image_surface_create_for_data()

**Comment by ggarra13** [comment]
I will try modify subwindow.cxx tomorrow adding some additional code to reproduce the behavior I am seeing.  Today it is kind of late.

**Comment by ggarra13** [comment]
Oh well... I did it anyway.  The code does not use resizable() but uses custom handle() checks to check resizing on all margins.  

[subwindow.cxx.txt](https://github.com/user-attachments/files/15527062/subwindow.cxx.txt)

**Comment by ManoloFLTK** [comment]
OK. There was a remaining missing bit in Wayland support of interactive subwindow resize. I have now committed the modified Wayland code to FLTK master. Your modified `subwindow.cxx `runs now well with the last but unpatched FLTK code. I would guess your app will run OK too with it. 

It's good you've made me complete the FLTK support of interactive moving and resizing of subwindows under Wayland. No test app allows to check that, and I had not thought at all about exploring this feature.

**Comment by ManoloFLTK** [comment]
The video above is fast, it's difficult to follow what happens. I see at worse the GL scene only transiently above the right window part. That seems acceptable to me for a interactive resizing operation.

My understanding is that issue can be closed now.

**Comment by ggarra13** [comment]
> The video above is fast, it's difficult to follow what happens. I see at worse the GL scene only transiently above the right window part. 

It is indeed in general transient.  Only once in my tests the redraw left garbage.

> My understanding is that issue can be closed now.

I am indeed closing the issue.  If I manage to get a simple test, I'll post a new issue.  I also understand you guys want to get 1.4 out the door, so I am sorry for being a pest with these late bugs.

**Comment by ggarra13** [comment]
> It's good you've made me complete the FLTK support of interactive moving and resizing of subwindows under Wayland. No test app allows to check that, and I had not thought at all about exploring this feature.

I'm afraid your patch is still not fully working perfectly.  I cannot reproduce it with subwindow.cxx from this thread, but in **some** windows in my program I am still getting wrong values when resizing from the left:

[left_resizing.webm](https://github.com/fltk/fltk/assets/6072843/05f448a2-ffa0-4afb-8339-f04bcc5996da)

**Comment by ManoloFLTK** [comment]
Is the problem that you stretch the window to the left from its left side and you see its left side jump to the right?

**Comment by ManoloFLTK** [comment]
That should be good with the commit mentioned above.

**Comment by ggarra13** [comment]
No good.  It is much worse than before.  Now I get flickering when moving the window and when hiding an element in an Fl_Tile, it does not refresh properly one of the elements of the tile.

**Comment by ggarra13** [comment]
> Is the problem that you stretch the window to the left from its left side and you see its left side jump to the right?

No.  The issue (before your new commit), is that the window would jump way too much to the left, leaving the screen and leading to:

 Can't create Cairo surface with cairo_image_surface_create_for_data()

With your new commit, the window now flickers while dragging and it has refresh issues once I hide/unhide one of the groups in the Fl_Tile (the group that has a gl window).

**Comment by ggarra13** [comment]
Here's a video of what was happening before your latest commit:

[BadResize.webm](https://github.com/fltk/fltk/assets/6072843/d7ab7455-e079-4f5a-9fec-8b415a398e1a)

**Comment by ggarra13** [comment]
Here's a video with the most recent commit:

[BadMove.webm](https://github.com/fltk/fltk/assets/6072843/4cca8bf3-f208-4999-8539-9d3e85309191)

**Comment by ManoloFLTK** [comment]
OK. I would need, please, to know what happens with the new commit and the modified subwindow test app given above.

Also, could you try with Ubuntu 24.04, if possible?

**Comment by ggarra13** [comment]
> OK. I would need, please, to know what happens with the new commit and the modified subwindow test app given above.

Dragging seems fine, with no jittering.  After 3 rounds of resizing I got:

```
# gga@ubuntu22.04 ~/code/tests/fltk/wayland_subwindow_drag_resize 
# mié jun 12 05:27:21 [523] $ subwindow
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Abortado (`core' generado)

# gga@ubuntu22.04 ~/code/tests/fltk/wayland_subwindow_drag_resize 
# mié jun 12 05:27:21 [523] $ subwindow
Can't create Cairo surface with cairo_image_surface_create_for_data()

# gga@ubuntu22.04 ~/code/tests/fltk/wayland_subwindow_drag_resize 
# mié jun 12 05:28:46 [523] $ subwindow
Can't create Cairo surface with cairo_image_surface_create_for_data()
```

The bad_alloc issue I could not reproduce (it happened only once).  The Cairo surface error happened when I dragged the window to the right (without exiting the work area) and quickly dragged the left border to the left end of the screen (without, I believe, getting to exit it).

> Also, could you try with Ubuntu 24.04, if possible?

I'll see if I can install Virtual Box on my Linux machine and figure out how to install Ubuntu 24.04....

**Comment by ggarra13** [comment]
I am trying KVM on my Ubuntu 22.04.4 LTS box.  I got to install Ubuntu 24.04.4 LTS, but I am having a hard time getting a shared directory working on both machines.

**Comment by ggarra13** [comment]
Ok.  I tried both my program and subwindow.cxx on my Ubuntu 24.04.4 LTS with your latest patch and with Mesa drivers running under KVM.  

First, I did **NOT** recompile neither program under Ubuntu 24.04.4 LTS as KVM forces me to use nfs as file system which is **really** slow.

Without recompiling under Ubuntu 24.04.4 LTS, they both exhibit problems when dragging around.  In the case of subwindow.cxx, this can be seen while dragging it around really fast (it will work a while but then suddenly will flicker all over).
On my application, resizing from the left size would work better, albeit sometimes it would still jump to a new position.

I went ahead and **DID** recompile subwindow.cxx under Ubuntu's 24.04.4 LTS and dragging and resizing issues seemed to have been gone.  Resizing still exhibits only one problem.  If the dragging is taken to make it smaller than the window, the opposite side will actually expand (this can be seen when dragging from the left or from the top).

Rocky Linux 8.10 just updated its Wayland support, albeit I cannot get Wayland to work there with my NVidia graphic card.  I'll keep you posted.

**Comment by ManoloFLTK** [comment]
> I went ahead and DID recompile subwindow.cxx under Ubuntu's 24.04.4 LTS and dragging and resizing issues seemed to have been gone.

Thanks for doing this. These results suggest the problem is in the relatively young Wayland code in Ubuntu 22.04 and has been fixed in 24.04. I conclude FLTK needs not search for a fix that may just not exist. It would still be interesting to know what do you see with your recompiled program under 24.04.

**Comment by ggarra13** [comment]
> It would still be interesting to know what do you see with your recompiled program under 24.04.

I went ahead and upgraded to 24.04 (which was a royal pain as it broke things in the upgrade and I had to install fresh from the ISO).  I am recompiling my program as I type this.

**Comment by ggarra13** [comment]
Sigh :(  I stumbled onto a crashing error which shows up on Ubuntu 24.04.4 LTS.

I call Fl::flush before the Fl::run() loop to show up the main which was shown before.
With FLTK's master HEAD, I get the following stack trace on Wayland (XWayland is fine):

```
0x5ca78feabbca signal_callback
 (/home/gga/code/applications/mrv2/mrv2/lib/mrvCore/Linux/mrvSignalHandler.cpp:21)
0x7003efc3776e _dl_close
 (./elf/dl-close.c:770)
0x7003efc3651b __GI__dl_catch_exception
 (./elf/dl-catch.c:237)
0x7003efc36668 _dl_catch_error
 (./elf/dl-catch.c:256)
0x7003ec697c72 _dlerror_run
 (./dlfcn/dlerror.c:138)
0x7003ec6979a5 __dlclose
 (./dlfcn/dlclose.c:31)
0x5ca78fde22a1 Fl_Wayland_Gl_Window_Driver::set_gl_context(Fl_Window*, void*)
 (/home/gga/code/applications/mrv2/BUILD-Linux-amd64/Debug/FLTK-prefix/src/FLTK/src/drivers/Wayland/Fl_Wayland_Gl_Window_Driver.cxx:205)
0x5ca78fddb1da Fl_Gl_Window::make_current()
 (/home/gga/code/applications/mrv2/BUILD-Linux-amd64/Debug/FLTK-prefix/src/FLTK/src/Fl_Gl_Window.cxx:127)
0x5ca78fddb4a8 Fl_Gl_Window::flush()
 (/home/gga/code/applications/mrv2/BUILD-Linux-amd64/Debug/FLTK-prefix/src/FLTK/src/Fl_Gl_Window.cxx:205)
0x5ca78f990a56 Fl_Window_Driver::flush()
 (/home/gga/code/applications/mrv2/BUILD-Linux-amd64/Debug/FLTK-prefix/src/FLTK/src/Fl_Window_Driver.cxx:190)
0x5ca78f9d5948 Fl_Wayland_Window_Driver::flush()
 (/home/gga/code/applications/mrv2/BUILD-Linux-amd64/Debug/FLTK-prefix/src/FLTK/src/drivers/Wayland/Fl_Wayland_Window_Driver.cxx:401)
0x5ca78f95fb6e Fl::flush()
 (/home/gga/code/applications/mrv2/BUILD-Linux-amd64/Debug/FLTK-prefix/src/FLTK/src/Fl.cxx:773)
0x5ca78f8e8de2 mrv::App::run()
 (/home/gga/code/applications/mrv2/mrv2/lib/mrvApp/mrvApp.cpp:1061)
0x5ca78f8e1a8c main
 (/home/gga/code/applications/mrv2/mrv2/src/main.cpp:70)
```

**Comment by ggarra13** [comment]
Okay.  The problem seems to be with opengl windows.  None of the FLTK demo gl tests that have OpenGL in them seem to work for me under a fresh install of Ubuntu 24.04.4 LTS.

**Comment by ggarra13** [comment]
It is an Ubuntu 24.04.4/NVidia driver issue.  They completely broke Wayland.  NVidia driver 550 fixes it partially (FLTK demos will work) but any application creating textures or switching contexts between threads does not work.

**Comment by ManoloFLTK** [comment]
> I call Fl::flush before the Fl::run() loop to show up the main window which was shown before.
With FLTK's master HEAD, I get the following stack trace on Wayland (XWayland is fine):

I believe you should call `Fl_Window()::wait_for_expose()` in that case.

**Comment by ManoloFLTK** [comment]
> It is an Ubuntu 24.04.4/NVidia driver issue. They completely broke Wayland. NVidia driver 550 fixes it partially (FLTK demos will work) but any application creating textures or switching contexts between threads does not work.

I don't understand how bugs appear with new versions rather than get fixed.

FLTK does use GL textures to draw text in GL scenes (e.g. test/cube). Does that work?

**Comment by ggarra13** [comment]
> I don't understand how bugs appear with new versions rather than get fixed.
> 

I don't understand it either.  Ubuntu 22.04.4 LTS was working close to perfect.
Now It seems NVidia is moving their Linux driver to be partially open source?
I am not sure I am getting HW acceleration, as I get the "dri2" message:

This is what I get with their drivers now in my OpenGL viewer:

Vendor:     Mesa
Renderer:   zink Vulkan 1.3(NVIDIA GeForce RTX 3080 (NVIDIA_PROPRIETARY))
Version:    4.6 (Compatibility Profile) Mesa 24.0.5-1ubuntu1

Max. Texture Size:16384 x 16384

EGL client apis:OpenGL OpenGL_ES 
EGL version:1.5
EGL vendor: Mesa Project
No display extensions:EGL_EXT_platform_base EGL_EXT_device_base EGL_EXT_device_enumeration EGL_EXT_device_query EGL_KHR_client_get_all_proc_addresses EGL_EXT_client_extensions EGL_KHR_debug EGL_KHR_platform_x11 EGL_EXT_platform_x11 EGL_EXT_platform_device EGL_MESA_platform_surfaceless EGL_EXT_explicit_device EGL_KHR_platform_gbm EGL_MESA_platform_gbm EGL_EXT_platform_wayland EGL_KHR_platform_wayland EGL_EXT_platform_xcb

I used to get this:

	Vendor:     NVIDIA Corporation
	Renderer:   NVIDIA GeForce RTX 3080/PCIe/SSE2
	Version:    4.6.0 NVIDIA 535.171.04

	Max. Texture Size:32768 x 32768
	
> FLTK does use GL textures to draw text in GL scenes (e.g. test/cube). Does that work?

No.  It hangs.  I'll report it to NVidia too.

**Comment by ManoloFLTK** [comment]
Overall, it seems that what concerns the Wayland source code of FLTK has been implemented and remaining issues are caused by the NVidia driver. That would suggest this issue can be closed.

**Comment by ggarra13** [comment]
Ok.  I found that NVidia had nothing to do it with it.   My first issue was with an update to glfw3 that broke Wayland support.

The second issue is a regression in FLTK with this tag:
d7d1d0d6b2be85d562d6c0a0d910e2af260f8a6b

With that commit, resizing of my main window in the viewer, leaves behind the other subwindow (timeline).   See attached picture:

![Wayland_subwindow_notresizing](https://github.com/fltk/fltk/assets/6072843/d07155a3-9849-4ce5-8f6a-111dcfc25850)

**Comment by ManoloFLTK** [comment]
Please, explain what is wrong when you resize your main window. The attached picture is not understandable for me.

**Comment by ggarra13** [comment]
The subwindow, that contains the Play, Stop, Step and frame controls is not placed at the bottom of the window, but left flying where the original window opened (before it got resized to the size of the movie file).

Here's a picture of how it *should* be and was before that commit:

![Wayland_ok](https://github.com/fltk/fltk/assets/6072843/4c85cbaa-8ffc-41e3-a195-a16318775cca)

**Comment by ggarra13** [comment]
@ManoloFLTK Note: The resize is done programatically, not through the resize borders.

**Comment by ManoloFLTK** [comment]
@ggarra13 I have reproduced the problem you report when resizing the main window just after its initial map on the display. The commit reported above fixes this problem. It also supports properly according to Wayland move and resize operations applied to a subwindow. Therefore, I believe we are good now.

**Comment by ggarra13** [comment]
@ManoloFLTK I am afraid there are still issues, mainly when hiding/unhiding of subwindows (or their parents) at least with Fl_Flex.  This can be seen in the following videos (you can see them with ffplay for actual speed):

X11 correct hiding/unhiding behavior:
[X11.webm](https://github.com/fltk/fltk/assets/6072843/c2032c46-c284-4c77-9e4e-4165866bdb98)

Wayland incorrect hiding/unhiding behavior:
[Wayland.webm](https://github.com/fltk/fltk/assets/6072843/e198babc-c0a5-47f4-83d0-2e47f183d3ac)

X11 correct hiding/unhiding of pixel bar during playback:
[X11_Timeline.webm](https://github.com/fltk/fltk/assets/6072843/37c154f2-b8fe-49ed-8e2b-b7f73bc352c0)

Wayland disappearing timeline (first a gap and when timeline's group is redrawn -- timeline disappears):
[Wayland_Timeline.webm](https://github.com/fltk/fltk/assets/6072843/3d649374-764e-44b0-9396-5c3e5e2d75d8)

**Comment by ggarra13** [comment]
@ManoloFLTK FYI, I checked tag b17357339f66949de7787e83aa9d21d613439170 and it did not suffer from the disappearing timeline bug, albeit dragging around the panel would lead to bogus numbers.

**Comment by ManoloFLTK** [comment]
> The second issue is a regression in FLTK with this tag:
https://github.com/fltk/fltk/commit/d7d1d0d6b2be85d562d6c0a0d910e2af260f8a6b
With that commit, resizing of my main window in the viewer, leaves behind the other subwindow (timeline).

Please, report whether this issue is fixed with 26ea55d.

About "incorrect hiding/unhiding behavior of dock". Does your code add a child to an Fl_Flex when the moved subwindow enters the main window from its right, and then hides this child when it enters the left half of the main window? Could you transfer to your modified subwindow.cxx the code that handles the Fl_Flex? No OpenGL scene necessary in this test code.

**Comment by ggarra13** [comment]
> Please, report whether this issue is fixed with [26ea55d](https://github.com/fltk/fltk/commit/26ea55d5a48f545404d56f489e41e82fd5e63c9e).
> 

That's the tag I was testing and that exhibited the wrong timeline behavior.

> About "incorrect hiding/unhiding behavior of dock". Does your code add a child to an Fl_Flex when the moved subwindow enters the main window from its right, and then hides this child when it enters the left half of the main window? 

With the timeline bug, I  hide and unhide a subwindow.

With the docking bug, I also don't add/remove any widgets.  I just unhide/hide them. 

> Could you transfer to your modified subwindow.cxx the code that handles the Fl_Flex? No OpenGL scene necessary in this test code.

It's tricky, as I rely on several auxiliary classes.

**Comment by ManoloFLTK** [comment]
> With that commit, resizing of my main window in the viewer, leaves behind the other subwindow (timeline).

I repeat: Please, report whether this issue (not another issue) is fixed with https://github.com/fltk/fltk/commit/26ea55d5a48f545404d56f489e41e82fd5e63c9e. Yes or no?

**Comment by ggarra13** [comment]
> > With that commit, resizing of my main window in the viewer, leaves behind the other subwindow (timeline).
> 
> I repeat: Please, report whether this issue (not another issue) is fixed with [26ea55d](https://github.com/fltk/fltk/commit/26ea55d5a48f545404d56f489e41e82fd5e63c9e). Yes or no?

Yes.  Clipping is fixed.

**Comment by ggarra13** [comment]
Here's a modified subwindow.cxx with docking that exhibits the erratic drag behavior with tag 26ea55d.

[subwindow.cxx.txt](https://github.com/user-attachments/files/16019185/subwindow.cxx.txt)

**Comment by ManoloFLTK** [comment]
Thanks for the new version of subwindow.cxx. I do see here there's much wobbling with it.

I attach my modification of subwindow done yesterday independently from yours to introduce an Fl_Flex with one or two children according to the horizontal position of the dockable window. Here, it works nicely without any wobbling with the present FLTK source code. I didn't search what is the significant difference betwen this and yours subwindow.cxx
 
[subwindowV2.cxx.txt](https://github.com/user-attachments/files/16024894/subwindowV2.cxx.txt)

**Comment by ggarra13** [comment]
> I attach my modification of subwindow done yesterday independently from yours to introduce an Fl_Flex.

Thank you Manolo.  I actually found my modification of subwindow had a typo in it which was the reason it was misbehaving.  However that's not my problem in my actual application.  I am still investigating.

**Comment by ManoloFLTK** [comment]
I have found that several changes in the FLTK source code are needed to fully support moving and resizing a subwindow inside its parent. I will soon commit that and post here a test program that allows to drag around a dockable window, to show and hide a dock area and to contain a (toy) GL scene inside an Fl_Flex group.

**Comment by ManoloFLTK** [comment]
I attach a modified version of the FLTK test program shape that does for Wayland much of the operations involved in docking in presence of a toy GL scene placed inside an Fl_Flex that alternates between one or two children.

Drag or enlarge the docking window to the left to see the dock site appear in the main window while the GL scene is resized to make place for the dock site.

To run correctly, this requires commit 1c6a0c1 of FLTK.

That should hopefully solve the issues reported in your post above dated June 27.

[shapeV2.cxx.txt](https://github.com/user-attachments/files/16040220/shapeV2.cxx.txt)

**Comment by ggarra13** [comment]
I have already been testing commit 1c6a0c1 with no much success.

The issues I saw with shapeV2.cxx:

1) The GL scene is not drawn on my Ubuntu 24.04.4 LTS box with NVidia drivers (I attach a picture).
2) There's a noticeable slow down and lag when dragging the subwindow near the dock sometimes.
3) You are hard-coding the size of dock bar, instead of using Fl_Flex's layout() function.  In my actual program I discovered that calling Fl_Flex's layout() function under Wayland would set the x() position of the dock to a coordinate outside the main window (that's why there was flickering in my my hiding/unhiding of the dock movie of June 27).

![Captura desde 2024-06-29 14-43-10](https://github.com/fltk/fltk/assets/6072843/4b1c5141-4d88-4319-92ac-72335349fdb2)

**Comment by ggarra13** [comment]
And here are the results of commit 1c6a0c1 with my program:

1) The timeline now positions itself properly even after resized (bug fixed).
2) The other controls do not.  I think this is the Fl_Flex issue with Wayland returning incorrect coordinates for hidden elements in both of these movie errors I still see.

I am attaching movies (with explanations):

Docking/Undocking:
[Wayland_dock_after_flex_layout.webm](https://github.com/fltk/fltk/assets/6072843/15323541-33c8-4ad6-9ef0-392ad0e2d99c)
[X11_dock_after_flex_layout.webm](https://github.com/fltk/fltk/assets/6072843/e72b7971-ca67-4727-832c-6381a74859f3)

[Wayland_flex_layout_of_pixel_and_timeline_bar.webm](https://github.com/fltk/fltk/assets/6072843/9eafe911-bcc9-4875-9506-6644e530b8ac)
[X11_flex_layout_of_pixel_and_timeline_bar.webm](https://github.com/fltk/fltk/assets/6072843/9dc5ebfc-dec3-4564-a8b6-289d3ad0d99d)

FLUID layout of Timeline control subwindow and Pixel bar control subwindow:
![fluid_diagram](https://github.com/fltk/fltk/assets/6072843/297f1e07-e388-44e9-a6e2-25df8fa89714)

**Comment by ggarra13** [comment]
@ManoloFLTK @Albrecht-S I am CCing Albrecht-S since he is more familiar with the Fl_Flex code.

I started debugging Fl_Flex under Wayland which is misbehaving and I discovered that the width is not updated inmediately after a resize.  Here's my modified Fl_Flex code with debugging info:

```
  for (int i = 0; i < nc; i++) {
    Fl_Widget *c = child(i);
    if (!c->visible())
      continue;

    if (hori) {
      if (fixed(c)) {
          std::cerr << i << ") fixed xp=" << xp << " yp="
                    << yp << std::endl;
        c->resize(xp, yp, c->w(), hh);
      } else {
          std::cerr << i << ") flexible xp=" << xp << " yp="
                    << yp << " w(sp)=" << sp << std::endl;
        c->resize(xp, yp, sp, hh);
        if (--rem == 0) sp--;
      }
      std::cerr << i << ") xp CHECK=" << xp << " c->w()="
                << c->w() << " gap=" << gap_ << std::endl;
      
      xp += c->w() + gap_;
      
      std::cerr << i << ") xp NOW=" << xp << std::endl;
    } else {
      if (fixed(c)) {
        c->resize(xp, yp, vw, c->h());
      } else {
        c->resize(xp, yp, vw, sp);
        if (--rem == 0) sp--;
      }
      yp += c->h() + gap_;
    }
  }
```

The key issue is the value of c->w() after a flexible resize is not updated immediately and returns its previous size before the resize.  Here's the output I get:

```
0) fixed xp=0 yp=53
0) xp CHECK=0 c->w()=45 gap=0
0) xp NOW=45
1) flexible xp=45 yp=53 w(sp)=375
1) xp CHECK=45 c->w()=645 gap=0
1) xp NOW=690
2) fixed xp=690 yp=53
2) xp CHECK=690 c->w()=270 gap=0
2) xp NOW=960
```
Notice how w(sp) is set to 375 on the flexible widget (The GL Window) yet xp CHECK and xp NOW after updating updates with c->w()=643 instead.  Is there a round trip under Wayland that needs to happen so that the GL Window updates its sizes, Manolo?

**Comment by ggarra13** [comment]
With this patch to Fl_Flex, the docking behavior under Wayland is fixed in my program.  However I don't think this is a correct fix, but a work-around of a Wayland bug.

[fl_flex_v1.patch.txt](https://github.com/user-attachments/files/16042078/fl_flex_v1.patch.txt)

**Comment by ManoloFLTK** [comment]
> The issues I saw with shapeV2.cxx:
    1. The GL scene is not drawn on my Ubuntu 24.04.4 LTS box with NVidia drivers (I attach a picture).
   
What happens if you type ctrl-+ ? What happens if you resize the main window ? What happens if you drag the subwindow at right to its left until it enters the main window ? Does the GL scene appear after these actions ?

What happens if you run FLTK's unchanged subwindow program: does the GL scene appear ?

What happens if you run FLTK's unchanged subwindow program with FLTK at commit 216ddd0: does the GL scene appear ?

**Comment by ManoloFLTK** [comment]
> I am attaching movies (with explanations)

Sorry, that's useless for me. I understand nothing of all that.

**Comment by ManoloFLTK** [comment]
> You are hard-coding the size of dock bar, instead of using Fl_Flex's layout() function.

OK. I see this point. I attach shapeV3.cxx.txt that fixes this

[shapeV3.cxx.txt](https://github.com/user-attachments/files/16043347/shapeV3.cxx.txt)

**Comment by ManoloFLTK** [comment]
>  Is there a round trip under Wayland that needs to happen so that the GL Window updates its sizes, Manolo?

Test program shapeV3.cxx (or V2) tests resize operations inside the Fl_Flex, but not resizing the Fl_Flex itself inside the main window. There might be a missing thing in commit 1c6a0c1. Please, try applying the attached small patch to FLTK. It should render your fl_flex_v1.patch un-necessary.

[resize-Fl_Flex.patch](https://github.com/user-attachments/files/16043440/resize-Fl_Flex.patch)

**Comment by ggarra13** [comment]
> > The issues I saw with shapeV2.cxx:
> > 1. The GL scene is not drawn on my Ubuntu 24.04.4 LTS box with NVidia drivers (I attach a picture).
> 
> What happens if you type ctrl-+ ? 

The screen enlarges, the OpenGL does not show up.

> What happens if you resize the main window ? 

The black (GL?) window enlarges and the outside floating window follows along on the height, but not on the size or position.

> What happens if you drag the subwindow at right to its left until it enters the main window ? Does the GL scene appear after these actions ?
> 

Nope.  The dock bar shows up, but the gl window does not appear.

> What happens if you run FLTK's unchanged subwindow program: does the GL scene appear ?
> 

What subwindow program with GL?  Only shapeV2.cxx has a GL window.

> What happens if you run FLTK's unchanged subwindow program with FLTK at commit [216ddd0](https://github.com/fltk/fltk/commit/216ddd09c9255597ddb3a4cc8005dc94ad6878c1): does the GL scene appear ?

It never opens a window and hangs.  It just shows the icon on the dockbar.

**Comment by ggarra13** [comment]
> Please, try applying the attached small patch to FLTK. It should render your fl_flex_v1.patch un-necessary.
> 
> [resize-Fl_Flex.patch](https://github.com/user-attachments/files/16043440/resize-Fl_Flex.patch)

The patch does not seem to apply cleanly.  It complained with:

patching file FL/Fl_Group.H
patching file src/Fl_Flex.cxx
patch unexpectedly ends in middle of line

I tried it anyway and the x() position in the layout kept being incorrect.

**Comment by ManoloFLTK** [comment]
>>     What happens if you run FLTK's unchanged subwindow program: does the GL scene appear ?
>What subwindow program with GL? Only shapeV2/3.cxx has a GL window.

Sorry. I meant "What happens if you run FLTK's unchanged **shape** program: does the GL scene appear ?"

**Comment by ManoloFLTK** [comment]
> I tried it anyway and the x() position in the layout for the dock kept being incorrect.

I just tried here to apply this patch to a clean, updated, FLTK source code. It applies cleanly with command `patch -p1 < patch-file-name` :
```
git pull
patch -p1 < /Users/mgouy/VBoxShared/linuxshared/fltk-macshared/resize-Fl_Flex.patch 
patching file 'FL/Fl_Group.H'
patching file 'src/Fl_Flex.cxx'
```
Could you, please, try again?

**Comment by ggarra13** [comment]
# gga@Ubuntu 24.04 LTS ~/code/lib/fltk (master)
# dom jun 30 04:19:37 [639] $ git switch master -f
Ya en 'master'
Tu rama está actualizada con 'origin/master'.

# gga@Ubuntu 24.04 LTS ~/code/lib/fltk (master)
# dom jun 30 04:19:45 [640] $ git pull
Ya está actualizado.

# gga@Ubuntu 24.04 LTS ~/code/lib/fltk (master)
# dom jun 30 04:19:53 [641] $ patch -p1 < resize-Fl_Flex.patch 
patching file FL/Fl_Group.H
patching file src/Fl_Flex.cxx
patch unexpectedly ends in middle of line

**Comment by ggarra13** [comment]
> Sorry. I meant "What happens if you run FLTK's unchanged **shape** program: does the GL scene appear ?"

No, it does not.

**Comment by ManoloFLTK** [comment]
About `resize-Fl_Flex.patch`: there's a small mystery. Anyway, the important thing is to have 2 changes relatively to the FLTK of the git repo:
1) in file `FL/Fl_Group.H`, static int group_resize_depth_; must be in the protected instead of the private section of the class. 
2) in file `src/Fl_Flex.cxx`, function Fl_Flex::resize() should become
```
void Fl_Flex::resize(int x, int y, int w, int h) {
  bool is_resize = (w != this->w() || h != this->h());
  if (is_resize) group_resize_depth_++;
  Fl_Widget::resize(x, y, w, h);
  layout();
  if (is_resize) group_resize_depth_--;
} // resize()
```


> >Sorry. I meant "What happens if you run FLTK's unchanged shape program: does the GL scene appear ?"
> No, it does not. But "cube" works fine.

We have to keep separated two issues: how Fl_Flex resizes and how to map a GL subwindow with the NVidia driver.

1) please, try replacing the GL subwindow in shapeV3.cxx by a non-GL subwindow containing an Fl_Box with the label "GL here". Then run shapeV3 and report whether windows and subwindows and Fl_Flex behave correctly with `resize-Fl_Flex.patch`.

2) Can you find a way to map a GL subwindow correctly on screen? Your program does have several GL subwindow and they get mapped correctly. Are they all constantly changing  as in cube? I think we had program shape working in the past with NVidia? I can't experiment here because I can't reproduce the problem.

**Comment by ggarra13** [comment]
> We have to keep separated two issues: how Fl_Flex resizes and how to map a GL subwindow with the NVidia driver.
> 
> 1. please, try replacing the GL subwindow in shapeV3.cxx by a non-GL subwindow containing an Fl_Box with the label "GL here". Then run shapeV3 and report whether windows and subwindows and Fl_Flex behave correctly with `resize-Fl_Flex.patch`.

The patch had the changes.  Fl_Flex seems to behave properly in the shapeV3.cpp without a GL window.  However, it does not behave correctly in my application.  The GL window uses old widths when layout() is called.  I noticed your patch did not modify anything in layout or GL windows that would change that behavior.

> 2. Can you find a way to map a GL subwindow correctly on screen? 

OpenGL3test maps properly and does not have a constantly updating drawing.  In shapeV3.cxx the window maps correctly (ie. clearing the GL screen with glClearColor does change the subwindow's color).  It is the glVertex, etc. drawing calls that don't work. Could it be you are not setting compatibility with OpenGL's 2 core profile?

> Your program does have several GL subwindow and they get mapped correctly. Are they all constantly changing  as in cube? 

Not really, but they *are* OpenGL3 and they use textures, like in cube and OpenGL3test.

> I think we had program shape working in the past with NVidia? I can't experiment here because I can't reproduce the problem.

The NVidia problem was with EGL and that is fixed in Ubuntu 24.04.4 LTS.

**Comment by Albrecht-S** [comment]
@ManoloFLTK In comment https://github.com/fltk/fltk/issues/987#issuecomment-2198482832 you wrote:
> 2. in file src/Fl_Flex.cxx, function Fl_Flex::resize() should become ...

This is a bad approach. Setting (incrementing) a static variable of `Fl_Group` inside a derived class and resetting (decrementing) it later is error prone: **_If this had to be done_**, then this would also apply to other subclasses derived by users and that's something we should really, really, really avoid.

Furthermore, this would not be an issue on all platforms except Wayland (if I understand the issue correctly), hence code developed on other platforms (by users and FLTK devs) would break when ported to or used on Wayland. That's IMO a no-go!

I noticed your recent commit when you introduced this variable and my gut feeling was _"that's not how it should be done"_, but since it appeared to be a Wayland-only thing - and only in a core widget - I didn't interfere. But now this seems to become even worse.

**There should be a better solution!**

**Comment by ggarra13** [comment]
> > Sorry. I meant "What happens if you run FLTK's unchanged **shape** program: does the GL scene appear ?"
> 
> No, it does not. But "cube" works fine.

I am attaching a modified shape.cxx that actually works fine for me with NVidia drivers.  The only change is that the drawing is done on the GL_BACK buffer.  My actual program also has that modification for Wayland, as I noticed that drawing on the GL_FRONT buffer would not update until the next refresh.  Not sure if it is an FLTK bug, NVidia bug or just the way Wayland works (I remember asking Google's Gemini that told me it was the way Wayland worked but I obviously don't trust it very much).

[shape.cxx.txt](https://github.com/user-attachments/files/16046322/shape.cxx.txt)

**Comment by ManoloFLTK** [comment]
>  The only change is that the drawing is done on the GL_BACK buffer. My actual program also has that modification for Wayland, as I noticed that drawing on the GL_FRONT buffer would not update until the next refresh.

That's very useful information, thanks. I may find a way out.

**Comment by ManoloFLTK** [comment]
About the new Fl_Group static variable counting the depth of group resize operations:

It was born because of a peculiar requirement of Wayland when moving or resizing subwindows: Wayland **requires** the window encompassing the resized or moved subwindow to be redrawn (in the Wayland sense, that is, to instruct the compositor to re-map its graphics content), in addition to the subwindow itself of course. This requirement is explicitly stated in the `wl_subsurface` chapter of the Wayland core protocol.

If such resize/move is done because some widget above the subwindow is resized, then, the encompassing window is redrawn anyway, so no additional redraw is needed. That's where the new static counter variable works: it gives a way to detect whether an additional operation is necessary.

Class Fl_Group has many derived classes. I've looked at all of them. Most call Fl_Group::resize() when they override their virtual resize() member.  Two don't: Fl_Flex and Fl_Grid. Therefore, these two classes require to increment/decrement the new static counter. Others are unlikely to contain subwindows (e.g., Fl_Text_Display). Overall, it should be documented that when Fl_Group is derived and its resize() member is overridden by an FLTK user or developer,  Fl_Group::resize() has to be called by the override or the counter has to be incremented/decremented. That's a situation for developers comparable, in my view, to the requirement to call Fl_Group::init_sizes() when children are rearranged.

This issue has arisen only very recently because subwindows had til now been resized only when their top-level window was. Until Gonzalo put forward the requirement for window docking, which in Wayland can be done with a subwindow that has to be moved or resized relatively to its non-resized parent window.

Another element of the issue is that Wayland requires to synchronize redraw operations with the compositor's ability to perform them. This happens when a (sub)window is resized or a subwindow is moved inside its parent. Wayland gives a way to check whether the compositor is ready for a new redraw of a particular (sub)window. When it's not, the redraw operation is not performed. Therefore, to move a subwindow requires to know whether its parent is ready for a redraw if it's not redrawn anyway by a previous operation.

Any other idea is welcome. As of today I've not found how to resize or move subwindows correctly under Wayland without the new counter variable.

**Comment by ManoloFLTK** [comment]
>>The only change is that the drawing is done on the GL_BACK buffer. My actual program also has that modification for Wayland, as I noticed that drawing on the GL_FRONT buffer would not update until the next refresh.

>That's very useful information, thanks. I may find a way out.

@ggarra13  Here is a patch that hopefully could fix how the unchanged FLTK shape test app starts with the NVidia driver. Could you, please, try it and report here what you get?

[fix-GL-map.patch.txt](https://github.com/user-attachments/files/16050004/fix-GL-map.patch.txt)

**Comment by ggarra13** [comment]
> @ggarra13 Here is a patch that hopefully could fix how the unchanged FLTK shape test app starts with the NVidia driver. Could you, please, try it and report here what you get?

Now I am getting black with or without glDrawBuffer(GL_BACK).

**Comment by ManoloFLTK** [comment]
> It is not glDrawBuffer(GL_BACK) that makes the NVidia driver work, but glReadBuffer(GL_BACK) at the end. 

Without access to a Linux box with an NVidia driver and without knowledge of OpenGl, I'm not able to solve this problem. Also, I'm very puzzled by the appearance of function `glReadBuffer()` to fix the shape test program because that function is documented as follows
> glReadBuffer specifies a color buffer as the source for subsequent glReadPixels, glCopyTexImage1D, glCopyTexImage2D, glCopyTexSubImage1D, glCopyTexSubImage2D, and glCopyTexSubImage3D commands.

But none of these functions is used in the shape program!

If you replace in shape.cxx the statement `glReadBuffer(GL_BACK);` by
```
static int count=0; if(++count == 1) glReadBuffer(GL_BACK);
```
so that `glReadBuffer()` is called only once, does that repair the shape program? If so, I could modify the FLTK Wayland source code so it calls `glReadBuffer(GL_BACK);` once when creating a GL subwindow.

**Comment by ggarra13** [comment]
> so that `glReadBuffer()` is called only once, does that repair the shape program?

It displays the triangle at the beginning, but as soon as you try the slider to change it to a circle, the display remains black.  The call has to be permanent not just once.

> But none of these functions is used in the shape program!

Right.  But where/how does Wayland know where to write and copy the buffer from?  And how does it inform Fl_Gl_Window about it? And viceversa.

Explanation about how OpenGL works on different OSes follows.

With OpenGL, under X11 and Windows, you can specify FL_DOUBLE or not.  

If FL_DOUBLE is set there are two buffers: the back buffer and the front buffer.  By default, with FLTK you draw to the back buffer and then FLTK is in charge of swapping the buffers by calling swap_buffers() once draw() returns.  

Under macOS, even if FL_DOUBLE is not set, FLTK's behavior is that it creates two buffers nevertheless (I found that by trial and error).  My understanding of Wayland (and from what I see now) is that it works like macOS too.

Under macOS and Wayland, it seems the default behavior, even without FL_DOUBLE, is to write to the back buffer that the Window manager creates.  FLTK's macOS implementation automatically copies the back buffer to the front buffer once draw() returns.

Wayland seems like it is not qute doing that.  With the glReadBuffer(GL_BACK) it seems like Wayland gets the hint to copy the back buffer it created with the NVidia drivers.  Without it, it is either not triggering the copying or just copying the front buffer instead.   This may be a bug in the NVidia driver or something missing on the FLTK implementation.  You probably know better.

**Comment by Albrecht-S** [comment]
> About the new Fl_Group static variable counting the depth of group resize operations ...

@ManoloFLTK I created GitHub **_Discussion_** #999 to separate this part of the issue from the main issue. I'm investigating and will follow-up with details there. This should be mostly independent of the main issue here.

**Comment by ManoloFLTK** [comment]
> Wayland seems like it is not qute doing that. With the glReadBuffer(GL_BACK) it seems like Wayland gets the hint to copy the back buffer it created with the NVidia drivers. Without it, it is either not triggering the copying or just copying the front buffer instead. This may be a bug in the NVidia driver or something missing on the FLTK implementation. 

The FLTK source code for Wayland adds the FL_DOUBLE flag to all Fl_Gl_Window's mode() if it's not present already. The net result is that all FLTK GL windows are double buffered under Wayland.

Could you, please, try to modify in file `src/drivers/Wayland/Fl_Wayland_Gl_Window_Driver.cxx`, the end of function `Fl_Wayland_Gl_Window_Driver::swap_buffers()` to obtain this
```
    }
    glReadBuffer(GL_BACK); // seems necessary with NVidia driver  <-- THIS LINE TO BE ADDED
    eglSwapBuffers(Fl_Wayland_Gl_Window_Driver::egl_display, egl_surface);
  }
```
and report whether FLTK and your GL-using apps work with NVidia with that. I have tried here this change and all GL apps tested that run well without it run also OK with it. If that makes apps work with NVidia under Wayland and if you tell me this change makes OpenGL sense, I will commit it.

> You probably know better.

I don't because I understand nearly nothing of OpenGL.

**Comment by ggarra13** [comment]
I think the redraw issue is an NVidia bug.  The best you can do is indeed keep the glReadBuffer(GL_BACK); which should be harmless anyway.

I want to recap where we are now:

1) Dragging, resizing and docking a subwindow with Fl_Flex works properly.
2) FLTK tests work properly on NVidia if you apply the patch and if we can live with one redraw behind.
3) With the subwindow change, there are now a lot of bugs regarding GL subwindows (I am opening a new issue).

If you are okay with 2), I am for closing this issue.

**Comment by ManoloFLTK** [comment]
> 1 Dragging, resizing and docking a subwindow with Fl_Flex works properly.
>  3  With the subwindow change, there are now a lot of bugs regarding GL subwindows (I am opening a new issue).

Points 1 and 3 above look contradictory to me. Does that mean that docking works OK only without any GL subwindow involved?


> FLTK tests work properly on NVidia if you apply the patch and if we can live with one redraw behind.

I think it's not appropriate to apply this patch considering the following facts:
- an NVIdia driver bug seems to create the problem, not an error in FLTK ;
- the patch doesn't solve completely the GL drawing problem because some redraws remain missed;
- I don't understand why the patch would improve FLTK.

**Comment by ggarra13** [comment]
> Points 1 and 3 above look contradictory to me. Does that mean that docking works OK only without any GL subwindow involved?

Docking of subwindows works properly now.  But subwindows that don't have the leak outside flag and start in the main window do not resize/reposition themselves properly upon resizing of the main window.  See the new issue #1003 I opened for a complete test suite of the problems that appear.

**Comment by ggarra13** [comment]
> I think it's not appropriate to apply this patch considering the following facts:

Seems fine to me.  The problem only shows in OpenGL 2 FLTK demos, not in OpenGL 3 code it seems.

**Comment by ManoloFLTK** [comment]
@ggarra13 Should this issue remain open now that #1003 is fixed and closed?

**Comment by ggarra13** [comment]
No, I am closing it.  Thanks for all your hard work, @ManoloFLTK.
